cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/spark/ExecutionSparkContext.java

```diff
@@ -113,10 +115,43 @@ public ExecutionSparkContext(SparkSpecification specification, Id.Program progra
 
   @Override
   public <T> T readFromDataset(String datasetName, Class<?> kClass, Class<?> vClass) {
-    Map<String, String> datasetArgs = RuntimeArguments.extractScope(Scope.DATASET, datasetName, getRuntimeArguments());
-
     // Clone the configuration since it's dataset specification and shouldn't affect the global hConf
     Configuration configuration = new Configuration(hConf);
+
+    // first try if it is InputFormatProvider
+    Map<String, String> dsArgs = RuntimeArguments.extractScope(Scope.DATASET, datasetName, getRuntimeArguments());
+    Dataset dataset = getDataset(datasetName, dsArgs);
+    if (dataset instanceof InputFormatProvider) {
+      // get the input format and its configuration from the dataset
+      String inputFormatName = ((InputFormatProvider) dataset).getInputFormatClassName();
+
+      // load the input format class
+      if (inputFormatName == null) {
+        throw new DatasetInstantiationException(
+          String.format("Dataset '%s' provided null as the input format class name", datasetName));
+      }
+      try {
+        @SuppressWarnings("unchecked")
+        Class<? extends InputFormat> inputFormatClass = (Class<? extends InputFormat>) Class.forName(inputFormatName);
```

Comment:
Need to load it through the `SparkClassLoader`, otherwise cannot load classes inside user program jar.